src/main/java/htsjdk/samtools/util/output/async/AsyncBlockCompressedOutputStream.java

```diff
@@ -0,0 +1,238 @@
+package htsjdk.samtools.util.output.async;
+
+import htsjdk.samtools.util.BlockCompressedStreamConstants;
+import htsjdk.samtools.util.RuntimeIOException;
+import htsjdk.samtools.util.output.AbstractBlockCompressedOutputStream;
+import htsjdk.samtools.util.zip.DeflaterFactory;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.*;
+import java.util.function.Supplier;
+import java.util.zip.CRC32;
+import java.util.zip.Deflater;
+
+public class AsyncBlockCompressedOutputStream extends AbstractBlockCompressedOutputStream {
+
+    private static final int COMPRESSING_THREADS = Runtime.getRuntime().availableProcessors() - 2;
+
+    private static final ExecutorService gzipExecutorService = Executors.newFixedThreadPool(COMPRESSING_THREADS, runnable -> {
+        Thread thread = Executors.defaultThreadFactory().newThread(runnable);
+        thread.setDaemon(true);
+        return thread;
+    });
+
+    private static final Void NOTHING = null;
+    private static final int BLOCKS_PACK_SIZE = 16 * COMPRESSING_THREADS;
+
+    private DeflaterFactory deflaterFactory;
+    private final int compressionLevel;
+
+    private List<Future<CompressedBlock>> compressedBlocksInFuture = new ArrayList<>(BLOCKS_PACK_SIZE);
+    private CompletableFuture<Void> writeBlocksTask = CompletableFuture.completedFuture(NOTHING);
```

Comment:
The documentation of CompletableFuture says: 
> All async methods without an explicit Executor argument are performed using the ForkJoinPool.commonPool() (unless it does not support a parallelism level of at least two, in which case, a new Thread is created to run each task).

That's why I have to use `ExecutorService gzipExecutorService = Executors.newFixedThreadPool(COMPRESSING_THREADS, runnable -> {...});`, but it's still can be dangerous in case of the dual-core machine.
So, nice to have, we need to check the potential memory-leak situation somehow.

Reply:
Also it looks like non-managed resource. Thread pool is created once class is loaded and all threads will be around till jvm will shutdown (graceful shutdown also missing in class). Assume to look on cached thread pool.